libkmod/libkmod-builtin.c

```diff
@@ -1,333 +1,193 @@
-			void *tmp;
-			size_t s;
+	ssize_t len;
 
-			s = linesz + partsz;
-			tmp = realloc(iter->buf, s);
-
-			if (!tmp) {
-				sv_errno = errno;
-				goto fail;
-			}
-			iter->bufsz = s;
-			iter->buf = tmp;
-		}
-
-		strncpy(iter->buf + linesz, buf, partsz);
-		linesz += partsz;
+	len = getdelim(&info->buf, &info->bufsz, '\0', info->fp);
+	if (len > 0 && info->buf[len] != '\0') {
+		errno = EINVAL;
+		len = -1;
 	}
 
-	if (linesz) {
-		if (iter->buf[linesz - 1] != '\0') {
-			sv_errno = EINVAL;
-			goto fail;
-		}
-		*line = iter->buf;
-		*size = linesz;
-	}
-
-	return offset;
-fail:
-	errno = sv_errno;
-	return -1;
+	return len;
 }
 
-static bool kmod_builtin_iter_next(struct kmod_builtin_iter *iter)
+static ssize_t get_strings(struct kmod_builtin_info *info, const char *modname,
+				struct strbuf *buf)
 {
-	char *line,  *modname;
-	size_t linesz;
-	off_t pos, offset, modlen;
-
-	modname = NULL;
+	const size_t modlen = strlen(modname);
+	ssize_t count, n;
 
-	iter->nstrings = 0;
-	offset = pos = iter->next;
+	for (count = 0; count < INTPTR_MAX;) {
+		char *dot, *line;
 
-	while (offset < iter->size) {
-		char *dot;
-		off_t len;
-
-		offset = get_string(iter, pos, &line, &linesz);
-		if (offset <= 0) {
-			if (offset)
-				ERR(iter->ctx, "get_string: %s\n", strerror(errno));
-			pos = iter->size;
+		n = get_string(info);
+		if (n == -1) {
+			if (!feof(info->fp)) {
+				count = -errno;
+				ERR(info->ctx, "get_string: %s\n", strerror(errno));
+			}
 			break;
 		}
 
+		line = info->buf;
 		dot = strchr(line, '.');
-		if (!dot) {
-			ERR(iter->ctx, "kmod_builtin_iter_next: unexpected string without modname prefix\n");
-			pos = iter->size;
-			break;
+		if (dot == NULL) {
+			count = -EINVAL;
+			ERR(info->ctx, "get_strings: "
+			    "unexpected string without modname prefix\n");
+			return count;
 		}
-
-		len = dot - line;
-
-		if (!modname) {
-			modname = strdup(line);
-			modlen = len;
-		} else if (modlen != len || strncmp(modname, line, len)) {
+		if (strncmp(line, modname, modlen) || line[modlen] != '.') {
+			/*
+			 * If no string matched modname yet, keep searching.
+			 * Otherwise this indicates that no further string will
+			 * match again.
+			 */
+			if (count == 0)
+				continue;
 			break;
 		}
-
-		iter->nstrings++;
-		pos = offset;
+		if (!strbuf_pushchars(buf, dot + 1) ||
+		    !strbuf_pushchar(buf, '\0')) {
+			count = -errno;
+			ERR(info->ctx, "get_strings: "
+			    "failed to append modinfo string\n");
+			return count;
+		}
+		count++;
 	}
 
-	iter->pos = iter->next;
-	iter->next = pos;
-
-	free(modname);
+	if (count == INTPTR_MAX) {
+		count = -ENOMEM;
+		ERR(info->ctx, "get_strings: "
+		    "too many modinfo strings encountered\n");
+		return count;
+	}
 
-	return (iter->pos < iter->size);
+	return count;
 }
 
-static bool kmod_builtin_iter_get_modname(struct kmod_builtin_iter *iter,
-				char modname[static PATH_MAX])
+static char **strbuf_to_vector(struct strbuf *buf, size_t count)
```

Comment:
this reminds me of a similar functions systemd has with the name "strv". I think we may eventually migrate this function to strbuf.h, but I'm fine with keep it where it is for now.